sim_pipeline/halos/halos_lens.py

```diff
@@ -0,0 +1,233 @@
+
+    def random_position(self):
+        """
+        Generates and returns random positions in the sky using a uniform distribution.
+
+        Returns
+        -------
+        px, py : float
+            The generated random x and y coordinates inside the skyarea in arcsec.
+        """
+        phi = 2 * np.pi * np.random.random()
+        upper_bound = np.sqrt(self.sky_area / np.pi)
+        random_radius = 3600 * np.random.uniform(0, upper_bound)
+        px = random_radius * np.cos(2 * phi)
+        py = random_radius * np.sin(2 * phi)
+        return px, py
+
+    def get_nfw_kwargs(self):
+        """
+        Returns the angle at scale radius, observed bending angle at the scale radius, and positions of the halos in
+        the lens plane from physical mass and concentration parameter of an NFW profile.
+
+        Returns
+        -------
+        Rs_angle, alpha_Rs, px, py : np.array
+            Rs_angle (angle at scale radius) (in units of arcsec)
+            alpha_Rs (observed bending angle at the scale radius) (in units of arcsec)
+            Arrays containing Rs_angle, alpha_Rs, and x and y positions of all the halos.
+        """
+        n_halos = self.n_halos
+        Rs_angle, alpha_Rs = np.empty(n_halos), np.empty(n_halos)
+        px, py = np.empty(n_halos), np.empty(n_halos)
+        c_200 = np.empty(n_halos)
+        for h in range(n_halos):
+            lens_cosmo = LensCosmo(z_lens=self.redshift_list[h], z_source=9999, cosmo=self.cosmo)
+            c_200[h] = concentration_from_mass(z=self.redshift_list[h], mass=self.mass_list[h])
+            Rs_angle_h, alpha_Rs_h = lens_cosmo.nfw_physical2angle(M=self.mass_list[h],
+                                                                   c=c_200[h])
+            px[h], py[h] = self.random_position()
+            Rs_angle[h] = Rs_angle_h
+            alpha_Rs[h] = alpha_Rs_h
+        # TODO: Also z_source
+        return Rs_angle, alpha_Rs, px, py
+
+    def get_halos_lens_kwargs(self):
+        """
+        Constructs and returns the list of keyword arguments for each halo to be used in the lens model for lenstronomy.
+
+        Returns
+        -------
+        kwargs_halos : list of dicts
+            The list of dictionaries containing the keyword arguments for each halo.
+        """
+        Rs_angle, alpha_Rs, px, py = self.get_nfw_kwargs()
+
+        kwargs_halos = [{'Rs': Rs_angle[h], 'alpha_Rs': alpha_Rs[h], 'center_x': px[h], 'center_y': py[h]}
+                        for h in range(self.n_halos)]
+        return kwargs_halos
+
+    def get_convergence_shear(self):
+        """
+        Calculates and returns the convergence and shear at the origin due to all the halos.
+
+        Returns
+        -------
+        kappa, gamma1, gamma2 : float
+            The computed convergence and two components of the shear at the origin.
+        """
+        kappa = self.lens_model.kappa(0.0, 0.0, self.get_halos_lens_kwargs(),
```

Comment:
you might be able to avoid duplicate calculations for kappa and gamma when you instead evaluate the hessian() and then calculate kappa and gamma values directly from it. Have a look at https://github.com/lenstronomy/lenstronomy/blob/e4c16d3b523c0f299aa57f2c0728565541e0258f/lenstronomy/LensModel/lens_model.py#L204
And https://github.com/lenstronomy/lenstronomy/blob/e4c16d3b523c0f299aa57f2c0728565541e0258f/lenstronomy/LensModel/lens_model.py#L247 , https://github.com/lenstronomy/lenstronomy/blob/e4c16d3b523c0f299aa57f2c0728565541e0258f/lenstronomy/LensModel/lens_model.py#L289